collect/multi_loop_test.go

```diff
@@ -0,0 +1,602 @@
+		return float64(totalSent) >= float64(len(allSpans))*0.45
+	}, 2*time.Second, transmission.WaitTime, "Should have sent some spans to transmission")
+}
+
+// TestMemoryPressureWithConcurrency tests the collector's behavior under memory pressure
+// with concurrent operations. This seems like a silly test in that it doesn't assert
+// much, but it can reveal deadlocks or races around checkAlloc.
+func TestMemoryPressureWithConcurrency(t *testing.T) {
+	const (
+		numTraces     = 500
+		spansPerTrace = 100
+		maxAlloc      = 10 * 1024 * 1024 // 10MB limit
+	)
+
+	conf := &config.MockConfig{
+		GetTracesConfigVal: config.TracesConfig{
+			SendTicker:   config.Duration(100 * time.Millisecond),
+			SendDelay:    config.Duration(50 * time.Millisecond),
+			TraceTimeout: config.Duration(2 * time.Hour),
+			MaxBatchSize: 500,
+		},
+		GetCollectionConfigVal: config.CollectionConfig{
+			CacheCapacity: 5000,
+			MaxAlloc:      config.MemorySize(maxAlloc),
+		},
+		GetSamplerTypeVal:  &config.DeterministicSamplerConfig{SampleRate: 1},
+		ParentIdFieldNames: []string{"trace.parent_id", "parentId"},
+		TraceIdFieldNames:  []string{"trace.trace_id", "traceId"},
+		SampleCache: config.SampleCacheConfig{
+			KeptSize:          100,
+			DroppedSize:       100,
+			SizeCheckInterval: config.Duration(1 * time.Second),
+		},
+	}
+
+	clock := clockwork.NewFakeClock()
+	collector := newTestCollector(t, conf, clock)
+	transmission := collector.Transmission.(*transmit.MockTransmission)
+
+	var wg sync.WaitGroup
+	spansAdded := int32(0)
+
+	// Drain the transmission channel; we're sending a lot of events here and we
+	// don't want it to back up.
+	go func() {
+		// This will terminate after we stop transmission and close the channel.
+		for range transmission.Events {
+		}
+	}()
+
+	// Add spans concurrently to trigger memory pressure
+	numGoroutines := 5
+	for g := 0; g < numGoroutines; g++ {
+		wg.Add(1)
+		go func(id int) {
+			defer wg.Done()
+
+			for i := 0; i < numTraces/numGoroutines; i++ {
+				traceID := fmt.Sprintf("trace-%d-%d", id, i)
+
+				for j := 0; j < spansPerTrace; j++ {
+					span := &types.Span{
+						Event: types.Event{
+							APIHost:    "http://api.honeycomb.io",
+							APIKey:     legacyAPIKey,
+							Dataset:    "test.dataset",
+							SampleRate: 1,
+							Timestamp:  time.Now(),
+							Data:       types.Payload{},
+						},
+						TraceID:     traceID,
+						IsRoot:      j == spansPerTrace-1,
+						ArrivalTime: time.Now(),
+					}
+
+					// Add large data to trigger memory pressure
+					span.Data.Set("large_field", make([]byte, 1024)) // 1KB per span
+
+					if err := collector.AddSpan(span); err == nil {
+						atomic.AddInt32(&spansAdded, 1)
+					}
+
+					clock.Advance(10 * time.Millisecond)
+				}
+			}
+		}(g)
+	}
+
+	wg.Wait()
+
+	// Verify the collector handled memory pressure
+	assert.Greater(t, atomic.LoadInt32(&spansAdded), int32(0), "Should have added spans")
```

Comment:
Maybe we can assert the `collector_cache_eviction` metric to at least check that the `checkAlloc` code has been triggered